File: CPP/No814.cc
/**
 * Created by Xiaozhong on 2020/12/30.
 * Copyright (c) 2020/12/30 Xiaozhong. All rights reserved.
 */

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode *pruneTree(TreeNode *root) {
        return containsOne(root) ? root : nullptr;
    }

private:
    bool containsOne(TreeNode *root) {
        if (root == nullptr) return false;
        bool a1 = containsOne(root->left);
        bool a2 = containsOne(root->right);
        if (!a1) root->left = nullptr;
        if (!a2) root->right = nullptr;
        return root->val == 1 || a1 || a2;
    }
};

File: Java/src/No814.java
import java.lang.reflect.InvocationTargetException;

/**
 * Java
 *
 * @Description: TODO
 * @Author: cherry
 * @Create on: 2022/7/21
 **/

public class No814 {
    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public TreeNode pruneTree(TreeNode root) {
        if (root == null) return null;
        boolean ret = dfs(root);
        return ret ? root : null;
    }

    private boolean dfs(TreeNode root) {
        if (root == null) return true;
        boolean left = dfs(root.left);
        boolean right = dfs(root.right);
        if (!left) root.left = null;
        if (!right) root.right = null;
        return root.val == 1 || left || right;
    }

    private static class B {
        // 删了之后抛异常 NoSuchMethodException
        public void hello() {
            System.out.println("hello");
        }

        public void hello(String name) {
            System.out.println(name);
        }
    }

    public static void main(String[] args) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
        B b = new B();
        b.getClass().getMethod("hello").invoke(b); // hello


        String s1 = "Hel" + new String("lo");
        String s2 = s1.intern();
        String s3 = "Hello";
        String s4 = "Hel" + "lo";
        String s5 = new String("Hello");
        String s6 = s5.intern();
        String s7 = "H";
        String s8 = "ello";
        String s9 = s7 + s8;
        final String s10 = "H";
        final String s11 = "ello";
        String s12 = s10 + s11;
        //jdk1.7+结果
        System.out.println(s3 == s1);// true
        System.out.println(s3 == s2);// true
        System.out.println(s3 == s4);// true
        System.out.println(s3 == s5);// false
        System.out.println(s3 == s6);// true
        System.out.println(s3 == s9);// false
        System.out.println(s3 == s12);// true
    }
}
